subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformExecutionBuildOperationIntegrationTest.groovy

```diff
@@ -176,6 +205,371 @@ class ArtifactTransformExecutionBuildOperationIntegrationTest extends AbstractIn
         }
     }
 
+    def "cacheability information for artifact transform executions is captured"() {
+        setupBuildWithColorTransform()
+        setupExternalDependency()
+        buildFile << """
+            @CacheableTransform
+            abstract class MakeGreen implements TransformAction<TransformParameters.None> {
+                @PathSensitive(PathSensitivity.RELATIVE)
+                @IgnoreEmptyDirectories
+                @NormalizeLineEndings
+                @InputArtifact
+                abstract Provider<FileSystemLocation> getInputArtifact()
+
+                @Classpath
+                @InputArtifactDependencies
+                abstract FileCollection getInputArtifactDependencies()
+
+                void transform(TransformOutputs outputs) {
+                    def input = inputArtifact.get().asFile
+                    println "processing [\${input.name}]"
+                    assert input.file
+                    def output = outputs.file(input.name + ".green")
+                    if (input.file) {
+                        output.text = input.text + ".green"
+                    } else {
+                        output.text = "missing.green"
+                    }
+                }
+            }
+        """
+
+        buildFile << """
+            project(":consumer") {
+                dependencies {
+                    implementation project(":producer")
+
+                    artifactTypes {
+                        jar {
+                            attributes.attribute(color, 'blue')
+                        }
+                    }
+                }
+            }
+        """
+        def producerTransformSpec = { Map<String, Object> it -> it.artifactName == 'producer.jar' }
+
+        when:
+        withBuildCache().run ":consumer:resolve"
+        then:
+        def projectExecution = findTransformExecution(producerTransformSpec)
+        with(snapshotInputsOperation(projectExecution).result) {
+            hash != null
+            classLoaderHash != null
+            implementationClassName == 'MakeGreen'
+            inputValueHashes.keySet() ==~ ['inputArtifactPath', 'inputPropertiesHash']
+            outputPropertyNames == ['outputDirectory', 'resultsFile']
+            inputFileProperties.keySet() ==~ ['inputArtifact', 'inputArtifactDependencies']
+            with(inputFileProperties.inputArtifactDependencies) {
+                hash != null
+                attributes == ['DIRECTORY_SENSITIVITY_DEFAULT', 'FINGERPRINTING_STRATEGY_CLASSPATH', 'LINE_ENDING_SENSITIVITY_DEFAULT']
+                with(Iterables.getOnlyElement(roots)) {
```

Comment:
💭  This guava helper is used in a bunch of places already, and I personally don't like the way it reads.

Would you be up for introducing a helper method on our base spec that does something like this? I generally like the approach GE has taken with the `unique` function, but I would rather call it `expectSingle`.

Reply:
What I like about Kotlin is that extension functions there are discoverable. Maybe we could come up with an extension function instead, so we don't have to remember the name upfront? `roots.onlyElement()` -- also reads well, IMO